libs/nio/src/main/java/org/elasticsearch/nio/SocketChannelContext.java

```diff
@@ -288,4 +314,42 @@ protected int flushToChannel(FlushOperation flushOperation) throws IOException {
         }
         return totalBytesFlushed;
     }
+
+    private void configureSocket(Socket socket, boolean isConnectComplete) throws IOException {
+        if (socketOptionsSet) {
+            return;
+        }
+
+        try {
+            // Set reuse address first as it must be set before a bind call. Some implementations throw
+            // exceptions on other socket options if the channel is not connected. But setting reuse first,
+            // we ensure that it is properly set before any bind attempt.
+            socket.setReuseAddress(socketConfig.tcpReuseAddress());
+            socket.setKeepAlive(socketConfig.tcpKeepAlive());
+            socket.setTcpNoDelay(socketConfig.tcpNoDelay());
+            int tcpSendBufferSize = socketConfig.tcpSendBufferSize();
+            if (tcpSendBufferSize > 0) {
+                socket.setSendBufferSize(tcpSendBufferSize);
+            }
+            int tcpReceiveBufferSize = socketConfig.tcpReceiveBufferSize();
+            if (tcpReceiveBufferSize > 0) {
+                socket.setReceiveBufferSize(tcpReceiveBufferSize);
+            }
+            socketOptionsSet = true;
+        } catch (IOException e) {
```

Comment:
can you add trace logging here so that we debug possible problems with socket options being  set later?

Reply:
I don't think we can do that because we have no logging infra in the nio module.